README.md
# NAME

Prancer::Plugin::Database

# SYNOPSIS

This plugin enables connections to a database and exports a keyword to access
those configured connections.

It's important to remember that when running your application in a single-
threaded, single-process application server like, say, [Twiggy](https://metacpan.org/pod/Twiggy), all users of
your application will use the same database connection. If you are using
callbacks then this becomes very important and you will want to take care to
avoid crossing transactions or expecting a database connection or transaction
to be in the same state it was before a callback.

To use a database connector, add something like this to your configuration
file:

    database:
        connection-name:
            driver: Prancer::Plugin::Database::Driver::DriverName
            options:
                username: test
                password: test
                database: test
                hostname: localhost
                port: 5432
                autocommit: true
                charset: utf8
                connection_check_threshold: 10
                dsn_extra:
                    RaiseError: 0
                    PrintError: 1
                on_connect:
                    - SET search_path=public

The "connection-name" can be anything you want it to be. This will be used when
requesting a connection from the plugin to determine which connection to
return. If only one connection is configured and no specific connection name is
requested then the only configured connection will be returned. If more than
one connection is configured and no specific connection name is requested then
any connection that is named "default" will be used. Otherwise an error will be
thrown. For example:

    use Prancer::Plugin::Database qw(database);

    Prancer::Plugin::Database->load();

    # if only one connection is configured then this will return that
    # connection
    my $dbh = database;

    # if multiple connections are configured and one of them is called
    # "default" then this will return the connection named "default" or throw
    # an error if no connection is named "default"
    my $dbh = database;

    # this will return a configured connection named "foo" or throw an error if
    # no connections are named "foo"
    my $dbh = database("foo");

# OPTIONS

- database

    **REQUIRED** The name of the database to connect to.

- username

    The username to use when connecting. If this option is not set then the default
    is the user running the application server or the current user.

- password

    The password to use when connecting. If this option is not set then the default
    is to connect with no password.

- hostname

    The host name of the database server. If this option is not set then the
    default is to connect to localhost.

- port

    The port number on which the database server is listening. If this option is
    not set then the default is to connect on the database's default port.

- autocommit

    If set to a true value -- like 1, yes, or true -- then this will enable
    autocommit. If set to a false value -- like 0, no, or false -- then this will
    disable autocommit. By default, autocommit is enabled.

- charset

    The character set to connect to the database with. If this is set to "utf8"
    then the database connection will attempt to make UTF8 data Just Work if
    available.

- connection\_check\_threshold

    This sets the number of seconds that must elapse between calls to get a
    database handle before performing a check to ensure that a database connection
    still exists and will reconnect if one does not. This handles cases where the
    database handle hasn't been used in a while and the underlying connection has
    gone away. If this is not set then it will default to 30 seconds.

- dsn\_extra

    If you have any further connection parameters that need to be appended to the
    dsn then you can put them in the configuration as a hash. This hash will be
    merged into the default parameters and overwrite any that are duplicated. The
    dsn parameters set by default are `AutoCommit` to 1, `RaiseError` to 1, and
    `PrintError` to 0. This option will take precedence over the `autocommit`
    flag above.

- on\_connect

    This can be an array of commands execute on a successful connection. These will
    be executed on every connection so if the connection goes away but is re-
    established then these commands will be run again.

# CREDIT

This module is derived from [Dancer::Plugin::Database](https://metacpan.org/pod/Dancer::Plugin::Database). Thank you to David
Precious.

# COPYRIGHT

Copyright 2014, 2015 Paul Lockaby. All rights reserved.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# SEE ALSO

- [Prancer](https://metacpan.org/pod/Prancer)


Changes
Revision history for Perl extension Prancer::Plugin::Database

1.00 Sat Dec 13 05:45:34 UTC 2014
	- Initial release.

1.01 Wed Dec 24 04:50:14 UTC 2014
	- Fixing unit tests, example code, perl docs

1.02 Sun Feb  1 17:21:55 UTC 2015
	- Adding configuration option for setting additional dsn parameters.
	- Adding configuration option for running database statements on connection.

1.03 Sun Feb  8 06:45:16 UTC 2015
	- If no connection name given and only one connection exists then use it.


lib/Prancer/Plugin/Database/Driver.pm
package Prancer::Plugin::Database::Driver;

use strict;
use warnings FATAL => 'all';

use version;
our $VERSION = '1.03';

use Try::Tiny;
use Carp;

# even though this *should* work automatically, it was not
our @CARP_NOT = qw(Prancer Try::Tiny);

sub new {
    my ($class, $config, $connection) = @_;

    try {
        require DBI;
    } catch {
        my $error = (defined($_) ? $_ : "unknown");
        croak "could not initialize database connection '${connection}': could not load DBI: ${error}";
    };

    # this is the only required field
    unless ($config->{'database'}) {
        croak "could not initialize database connection '${connection}': no database name configured";
    }

    my $self = bless({}, $class);
    $self->{'_connection'}      = $connection;
    $self->{'_database'}        = $config->{'database'};
    $self->{'_username'}        = $config->{'username'};
    $self->{'_password'}        = $config->{'password'};
    $self->{'_hostname'}        = $config->{'hostname'};
    $self->{'_port'}            = $config->{'port'};
    $self->{'_autocommit'}      = $config->{'autocommit'};
    $self->{'_charset'}         = $config->{'charset'};
    $self->{'_check_threshold'} = $config->{'connection_check_threshold'} || 30;
    $self->{'_dsn_extra'}       = $config->{'dsn_extra'} || {};
    $self->{'_on_connect'}      = $config->{'on_connect'} || [];

    # store a pool of database connection handles
    $self->{'_handles'} = {};

    return $self;
}

sub handle {
    my $self = shift;

    # to be fork safe and thread safe, use a combination of the PID and TID (if
    # running with use threads) to make sure no two processes/threads share a
    # handle. implementation based on DBIx::Connector by David E. Wheeler
    my $pid_tid = $$;
    $pid_tid .= "_" . threads->tid if $INC{'threads.pm'};

    # see if we have a matching handle
    my $handle = $self->{'_handles'}->{$pid_tid} || undef;

    if ($handle->{'dbh'}) {
        if ($handle->{'dbh'}{'Active'} && $self->{'_check_threshold'} &&
            (time - $handle->{'last_connection_check'} < $self->{'_check_threshold'})) {

            # the handle has been checked recently so just return it
            return $handle->{'dbh'};
        } else {
            if ($self->_check_connection($handle->{'dbh'})) {
                $handle->{'last_connection_check'} = time;
                return $handle->{'dbh'};
            } else {
                # er need to reconnect
                carp "database connection to '${\$self->{'_connection'}}' went away -- reconnecting";

                # try to disconnect but don't care if it fails
                if ($handle->{'dbh'}) {
                    try { $handle->{'dbh'}->disconnect(); } catch {};
                }

                # try to connect again and save the new handle
                $handle->{'dbh'} = $self->_get_connection();
                return $handle->{'dbh'};
            }
        }
    } else {
        $handle->{'dbh'} = $self->_get_connection();
        if ($handle->{'dbh'}) {
            $handle->{'last_connection_check'} = time;
            $self->{'_handles'}->{$pid_tid} = $handle;
            return $handle->{'dbh'};
        }
    }

    return;
}

sub _get_connection {
    my $self = shift;

    my $dbh = undef;
    try {
        $dbh = DBI->connect(@{$self->{'_dsn'}}) || croak "${\$DBI::errstr}\n";

        # run any on_connect sql
        $dbh->do($_) for (@{$self->{'_on_connect'}});
    } catch {
        my $error = (defined($_) ? $_ : "unknown");
        croak "could not initialize database connection '${\$self->{'_connection'}}': ${error}";
    };

    return $dbh;
}

# Check the connection is alive
sub _check_connection {
    my $self = shift;
    my $dbh = shift;
    return 0 unless $dbh;

    if ($dbh->{Active} && (my $result = $dbh->ping())) {
        if (int($result)) {
            # DB driver itself claims all is OK, trust it:
            return 1;
        } else {
            # it was "0 but true" meaning the DBD doesn't implement ping and
            # instead we got the default DBI ping implementation. implement
            # our own basic check by performing a real simple query.
            return try {
                return $dbh->do("SELECT 1");
            } catch {
                return 0;
            };
        }
    }

    return 0;
}

# stolen from Hash::Merge::Simple
## no critic (ProhibitUnusedPrivateSubroutines)
sub _merge {
    my ($self, $left, @right) = @_;

    return $left unless @right;
    return $self->_merge($left, $self->_merge(@right)) if @right > 1;

    my ($right) = @right;
    my %merged = %{$left};

    for my $key (keys %{$right}) {
        my ($hr, $hl) = map { ref($_->{$key}) eq "HASH" } $right, $left;

        if ($hr and $hl) {
            $merged{$key} = $self->_merge($left->{$key}, $right->{$key});
        } else {
            $merged{$key} = $right->{$key};
        }
    }

    return \%merged;
}

1;
